drivers/ethernet/phy/phy_mii.c

```diff
@@ -405,87 +447,112 @@ static int phy_mii_cfg_link(const struct device *dev,
 
 	k_sem_take(&data->sem, K_FOREVER);
 
-	if (phy_mii_reg_read(dev, MII_ANAR, &anar_reg) < 0) {
-		ret = -EIO;
-		goto cfg_link_end;
-	}
-	anar_reg_old = anar_reg;
-
 	if (phy_mii_reg_read(dev, MII_BMCR, &bmcr_reg) < 0) {
 		ret = -EIO;
 		goto cfg_link_end;
 	}
+	bmcr_reg_old = bmcr_reg;
 
-	if (data->gigabit_supported) {
-		if (phy_mii_reg_read(dev, MII_1KTCR, &c1kt_reg) < 0) {
-			ret = -EIO;
+	if (adv_speeds & LINK_AUTO_NEGOTIATION_DISABLED) {
+		/* Disable auto-negotiation */
+		bmcr_reg &= ~(MII_BMCR_AUTONEG_ENABLE | MII_BMCR_SPEED_LSB | MII_BMCR_SPEED_MSB);
+
+		if (data->gigabit_supported && PHY_LINK_IS_SPEED_1000M(adv_speeds)) {
+			bmcr_reg |= MII_BMCR_SPEED_1000;
+		} else if (PHY_LINK_IS_SPEED_100M(adv_speeds)) {
+			bmcr_reg |= MII_BMCR_SPEED_100;
+		} else if (PHY_LINK_IS_SPEED_10M(adv_speeds)) {
+			bmcr_reg |= MII_BMCR_SPEED_10;
+		} else {
+			LOG_ERR("Invalid speed %d for PHY (%d)", adv_speeds, cfg->phy_addr);
+			ret = -EINVAL;
 			goto cfg_link_end;
 		}
-	}
 
-	if (adv_speeds & LINK_FULL_10BASE) {
-		anar_reg |= MII_ADVERTISE_10_FULL;
-	} else {
-		anar_reg &= ~MII_ADVERTISE_10_FULL;
-	}
-
-	if (adv_speeds & LINK_HALF_10BASE) {
-		anar_reg |= MII_ADVERTISE_10_HALF;
-	} else {
-		anar_reg &= ~MII_ADVERTISE_10_HALF;
-	}
-
-	if (adv_speeds & LINK_FULL_100BASE) {
-		anar_reg |= MII_ADVERTISE_100_FULL;
+		if (PHY_LINK_IS_FULL_DUPLEX(adv_speeds)) {
+			bmcr_reg |= MII_BMCR_DUPLEX_MODE;
+		} else {
+			bmcr_reg &= ~MII_BMCR_DUPLEX_MODE;
+		}
 	} else {
-		anar_reg &= ~MII_ADVERTISE_100_FULL;
-	}
+		if (phy_mii_reg_read(dev, MII_ANAR, &anar_reg) < 0) {
+			ret = -EIO;
+			goto cfg_link_end;
+		}
+		anar_reg_old = anar_reg;
 
-	if (adv_speeds & LINK_HALF_100BASE) {
-		anar_reg |= MII_ADVERTISE_100_HALF;
-	} else {
-		anar_reg &= ~MII_ADVERTISE_100_HALF;
-	}
+		if (adv_speeds & LINK_FULL_10BASE) {
```

Comment:
Same for the other checks

```suggestion
		if ((adv_speeds & LINK_FULL_10BASE) != 0U) {
```